radar/src/radar.rs

```diff
@@ -0,0 +1,603 @@
+    } else {
+        Ok(vec![])
+    }
+}
+
+/// Function to compute all changes to a contact list needed given a resources id, current position,
+/// radar receiver, all old contacts, and a map of all entity positions that are published.
+/// Changes are in the form of RadarContactDeltas, either specifying to Add, Remove, or Change a contact.
+fn radar_updates(
+    entity_id: &str,
+    current_position: &Position,
+    radar_receiver: &RadarReceiver,
+    old_contacts: &HashMap<String, RadarContact>,
+    all_positions: &HashMap<String, Position>,
+) -> Vec<RadarContactDelta> {
+    let mut deltas: Vec<RadarContactDelta> = Vec::new();
+    let contacts: Vec<String> = old_contacts
+        .values()
+        .map(|rc| rc.clone().entity_id)
+        .collect();
+    for (k, v) in all_positions.iter() {
+        if contacts.contains(k) {
+            let mut rid: String = "".to_string();
+            for (key, val) in old_contacts.iter() {
+                if val.entity_id == *k {
+                    rid = key.to_string();
+                    break;
+                }
+            }
+            if within_radius(current_position, v, radar_receiver.radius) {
+                let vector_to = current_position.vector_to(v);
+                deltas.push(RadarContactDelta::Change(
+                    rid.replace(":", "."),
+                    RadarContact {
+                        entity_id: k.clone().to_string(),
+                        distance: vector_to.mag,
+                        heading_xy: vector_to.heading_xy,
+                        heading_z: vector_to.heading_z,
+                    },
+                ));
+            } else {
+                deltas.push(RadarContactDelta::Remove(rid));
+            }
+        } else if entity_id != k && within_radius(current_position, &v, radar_receiver.radius) {
+            let vector_to = current_position.vector_to(v);
+            deltas.push(RadarContactDelta::Add(RadarContact {
+                entity_id: k.clone().to_string(),
+                distance: vector_to.mag,
+                heading_xy: vector_to.heading_xy,
+                heading_z: vector_to.heading_z,
+            }));
+        }
+    }
+    deltas
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+enum RadarContactDelta {
+    Add(RadarContact),
+    Remove(String),
+    Change(String, RadarContact),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct PositionValue {
+    values: Position,
+}
+
+/// Receives messages on the subject `event.decs.components.{shard}.{entity}.position.change`
+/// Stores entity position in-memory in the POSITIONS HashMap
+/// The cache is used later to discover nearby radar_contacts
+pub(crate) fn handle_entity_position_change(
+    _ctx: &CapabilitiesContext,
+    msg: messaging::BrokerMessage,
+) -> CallResult {
+    let subject: Vec<&str> = msg.subject.split('.').collect();
+    let position_value: PositionValue = serde_json::from_slice(&msg.body)?;
```

Comment:
This PositionValue struct was the way that I found out I could pull a value out of a JSON slice, but surely there's a way to just reference a field inside of it.

Reply:
you can create a `serde_json::Value` from a &str or a byte slice. once you have a `Value`, you can then index into it like a nested dictionary, e.g. `foo["values"]["x"]` or `foo["values"]["y"]`